auth-jwt-service/src/main/java/io/stargate/auth/jwt/AuthJwtService.java

```diff
@@ -69,6 +79,148 @@ public StoredCredentials validateToken(String token) throws UnauthorizedExceptio
+    JSONObject stargateClaims = extractClaimsFromJWT(token);
+
+    preCheckSchemaRead(stargateClaims, keyspace, table);
+
+    // TODO: [doug] 2020-11-2, Mon, 17:23 Finish implementing
+    return action.call();
+  }
+
+  @Override
+  public ResultSet executeSchemaWriteWithAuthorization(
+      Callable<ResultSet> action, String token, String keyspace, String table) throws Exception {
+    JSONObject stargateClaims = extractClaimsFromJWT(token);
+
+    preCheckSchemaWrite(stargateClaims, keyspace, table);
+
+    // Just return the result. No value in doing a post check since we can't roll back anyway.
+    return action.call();
+  }
+
+  private JSONObject extractClaimsFromJWT(String token) throws JSONException {
+    // Grab the custom claims from the JWT. It's safe to work with the JWT as a plain Base64 encoded
+    // json object here since by this point we've already authenticated the request.
+    String[] parts = token.split("\\.");
+    String decodedPayload = new String(Base64.getUrlDecoder().decode(parts[1]));
+    JSONObject payload = new JSONObject(decodedPayload);
+    return payload.getJSONObject(CLAIMS_FIELD);
+  }
+
+  private void preCheckSchemaRead(JSONObject stargateClaims, String keyspace, String table) {
+    // TODO: [doug] 2020-11-2, Mon, 17:23 Finish implementing
+  }
+
+  private void preCheckSchemaWrite(JSONObject stargateClaims, String keyspace, String table) {
+    // TODO: [doug] 2020-11-2, Mon, 17:23 Finish implementing
+  }
+
+  private void preCheckDataReadWrite(
+      JSONObject stargateClaims, List<String> primaryKeyValues, Table tableMetadata)
+      throws JSONException, UnauthorizedException {
+    List<Column> keys = tableMetadata.primaryKeyColumns();
+
+    if (primaryKeyValues.size() > keys.size()) {
+      throw new IllegalArgumentException("Provided more primary key values than exists");
+    }
+
+    for (int i = 0; i < primaryKeyValues.size(); i++) {
+      // If one of the columns exist as a field in the JWT claims and the values do not match then
+      // the request is not allowed.
+      if (stargateClaims.has(STARGATE_PREFIX + keys.get(i).name())) {
+
+        if (!Column.ofTypeText(keys.get(i).type())) {
+          throw new IllegalArgumentException(
+              "Column must be of type text to be used for authorization");
+        }
+
+        String stargateClaimValue = stargateClaims.getString(STARGATE_PREFIX + keys.get(i).name());
+        String columnValue = primaryKeyValues.get(i);
+        if (!stargateClaimValue.equals(columnValue)) {
+          throw new UnauthorizedException("Not allowed to access this resource");
+        }
+      }
+    }
+  }
+
+  private void postCheckDataRead(JSONObject stargateClaims, ResultSet resultSet)
+      throws JSONException, UnauthorizedException {
+    if (resultSet == null || resultSet.rows() == null) {
+      return;
+    }
+
+    // If scopes/fields line up with the resultset then return
+    for (Row row : resultSet.rows()) {
```

Comment:
I'm not sure how well this is going to work for clients. IIRC, the result set may not support re-iteration and if we iterate all the rows here, the higher level client may not 'see' any data... :thinking: It might be necessary to wrap the result set.

Reply:
Oh, good point. What about cloning?

Reply:
What do you mean by cloning? Mere object copies may not be enough at it ultimately links to paging state in C* :thinking:

Reply:
Oh so if I copy the resultSet and iterate over the copy it could still mess with the paging state of the original?

Reply:
It will not "break", I think, but I believe we cannot be sure we get exactly the same results.